### PR TITLE
CROM-6789 - read_lines of an empty string file produces non-empty Array output (2/2)

### PR DESCRIPTION
**Pull request by kpierre13**

### CODE REVIEW

File: centaur/src/main/resources/standardTestCases/file_to_lines/file_to_lines.wdl [hunk 1]
```diff
@@ -22,7 +25,7 @@ workflow wf_test {
     call file_to_lines
 
     output {
-        Int lines_null_count = length(file_to_lines.lines_null)
+        Int lines_null = length(file_to_lines.lines_null)
```
Comment (cjllanwarne) [inline centaur/src/main/resources/standardTestCases/file_to_lines/file_to_lines.wdl:28]: TBH I think I preferred the original name (this is the count, not the content)

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala [hunk 2]
```diff
@@ -56,6 +56,20 @@ object EngineFunctionEvaluators {
     Try(Await.result(ioFunctionSet.readFile(fileToRead.value, Option(sizeLimit), failOnOverflow = true), ReadWaitTimeout))
   }
 
+  def split_lines(read: String): List[String] = {
+    if (read.nonEmpty) {
+      if (read.matches("[\\n]+")) {
+        val empty_string_list = List("")
+        println(empty_string_list)
+        empty_string_list
```
Comment (cjllanwarne) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:64]: Remove `println` and then you can just immediately return the value

```suggestion
        List("")
```

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala [hunk 3]
```diff
@@ -56,6 +56,20 @@ object EngineFunctionEvaluators {
...
+      } else {
+        read.split(System.lineSeparator).toList
```
Comment (cjllanwarne) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:66]: Depending on how paranoid we are, we could potentially cover windows-style line endings too. Might become more relevant in the Azure world:
```suggestion
        read.split("\\r?\\n").toList
```

File: wdl/transforms/new-base/src/test/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluatorsSpec.scala [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  }
+  it should "return a three element array with three new line characters" in {
+    EngineFunctionEvaluators.split_lines("\n\n\n") shouldBe(List(""))
+  }
+  it should "return a three element array with three words and three double new line characters" in {
+    EngineFunctionEvaluators.split_lines("a word\n\na word\n\none more word\n\n") shouldBe(List("a word", "", "a word", "", "one more word"))
+  }
+//  it should "return an empty array with an empty file" in {
+//    EngineFunctionEvaluators.split_lines("") shouldBe(List.empty)
+//  }
```
Comment (cjllanwarne) [inline wdl/transforms/new-base/src/test/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluatorsSpec.scala:28]: Remove if we aren't going to use it

File: wdl/transforms/new-base/src/test/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluatorsSpec.scala [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package wdl.transforms.base.linking.expression.values
+
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+class EngineFunctionEvaluatorsSpec extends AnyFlatSpec with Matchers{
+  behavior of "read_lines"
+  it should "return an empty array with an empty file" in {
```
Comment (cjllanwarne) [inline wdl/transforms/new-base/src/test/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluatorsSpec.scala:8]: I would suggest replacing `with` with `from` throughout this test set

```suggestion
  it should "return an empty array from an empty file" in {
```

File: wdl/transforms/new-base/src/test/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluatorsSpec.scala [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  it should "return a one element array with a one line file" in {
+    EngineFunctionEvaluators.split_lines("one line") shouldBe(List("one line"))
+  }
+  it should "return a one element array with a one line file ending with a new line character" in {
+    EngineFunctionEvaluators.split_lines("one line\n") shouldBe(List("one line"))
+  }
+  it should "return a one element array with a single new line character" in {
+    EngineFunctionEvaluators.split_lines("\n") shouldBe(List(""))
+  }
+  it should "return a three element array with three new line characters" in {
```
Comment (cahrens) [inline wdl/transforms/new-base/src/test/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluatorsSpec.scala:20]: The description of this test case is not what the method actually does.

File: wdl/transforms/new-base/src/test/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluatorsSpec.scala [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    EngineFunctionEvaluators.split_lines("\n\n\n") shouldBe(List(""))
+  }
+  it should "return a three element array with three words and three double new line characters" in {
```
Comment (cahrens) [inline wdl/transforms/new-base/src/test/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluatorsSpec.scala:23]: The description of this test case is not what the method actually does.

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala [hunk 8]
```diff
@@ -56,6 +56,18 @@ object EngineFunctionEvaluators {
     Try(Await.result(ioFunctionSet.readFile(fileToRead.value, Option(sizeLimit), failOnOverflow = true), ReadWaitTimeout))
   }
 
+  def split_lines(read: String): List[String] = {
```
[the remark this answers is not in the document]
Comment (kpierre13) [reply wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:59]: There is a split function that the read_lines method used previously. Like mentioned in standup, this new split_lines function is more for the edge-case is which Daniel describes on the ticket. The previous split function did not allow for the specific splitting at new-line characters as this one does. This new function allows for splitting at new-line characters and also returning a List with empty quotes that are shown in an array. The previous did not show anything within the array.

File: wdl/transforms/new-base/src/test/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluatorsSpec.scala (hunk 6 above)
Comment (kpierre13) [reply wdl/transforms/new-base/src/test/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluatorsSpec.scala:20]: Is this referencing using the word 'array' as opposed to the word 'List'.

**Comment by cahrens** [comment]
We are not implementing this change at this time.

**Pull request closed without merging**